plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
 }
 
-func (p *PowerStat) addCurrentDramPowerConsumption(socketID string, acc telegraf.Accumulator) {
-	tags := map[string]string{
-		"package_id": socketID,
+// parsePackageRaplMetrics takes a slice of unique package metrics, and returns only the metrics which depend on rapl.
+func parsePackageRaplMetrics(metrics []string) []string {
+	slice := []string{
+		packageCurrentPowerConsumption.String(),
+		packageCurrentDramPowerConsumption.String(),
+		packageThermalDesignPower.String(),
 	}
 
-	fields := map[string]interface{}{
-		"current_dram_power_consumption_watts": roundFloatToNearestTwoDecimalPlaces(p.rapl.getRaplData()[socketID].dramCurrentEnergy),
+	packageRaplMetrics := make([]string, 0, len(slice))
+	for _, m := range metrics {
+		if choice.Contains(m, slice) {
+			packageRaplMetrics = append(packageRaplMetrics, m)
+		}
 	}
 
-	acc.AddGauge("powerstat_package", fields, tags)
+	return packageRaplMetrics
 }
 
-func (p *PowerStat) addPerCoreMetrics(acc telegraf.Accumulator) {
-	var wg sync.WaitGroup
-	wg.Add(len(p.msr.getCPUCoresData()))
+// parsePackageMsrMetrics takes a slice of unique package metrics, and returns only the metrics which depend on msr.
+func parsePackageMsrMetrics(metrics []string) []string {
+	slice := []string{
+		packageCPUBaseFrequency.String(),
+		packageTurboLimit.String(),
+	}
 
-	for cpuID := range p.msr.getCPUCoresData() {
-		go p.addMetricsForSingleCore(cpuID, acc, &wg)
+	packageMsrMetrics := make([]string, 0, len(slice))
+	for _, m := range metrics {
+		if choice.Contains(m, slice) {
+			packageMsrMetrics = append(packageMsrMetrics, m)
+		}
 	}
 
-	wg.Wait()
+	return packageMsrMetrics
 }
 
-func (p *PowerStat) addMetricsForSingleCore(cpuID string, acc telegraf.Accumulator, wg *sync.WaitGroup) {
-	defer wg.Done()
-
-	if p.cpuFrequency {
-		p.addCPUFrequencyMetric(cpuID, acc)
+// hasDuplicate takes a slice of a generic type, and returns true
+// if the slice contains duplicates. Otherwise, it returns false.
+func hasDuplicate[S ~[]E, E constraints.Ordered](s S) bool {
+	m := make(map[E]struct{}, len(s))
+	for _, v := range s {
+		if _, ok := m[v]; ok {
+			return true
+		}
+		m[v] = struct{}{}
 	}
+	return false
+}
 
-	// Read data from MSR only if required
-	if p.cpuC0StateResidency || p.cpuC1StateResidency || p.cpuC6StateResidency || p.cpuBusyCycles || p.cpuTemperature || p.cpuBusyFrequency {
-		err := p.msr.openAndReadMsr(cpuID)
+// parseCores takes a slice of strings where each string represents a group of
+// one or more CPU IDs (e.g. ["0", "1-3", "4,5,6"] or ["1-3,4"]). It returns a slice
+// of integers.
+func parseCores(cores []string) ([]int, error) {
+	parsedCores := make([]int, 0, len(cores))
+	for _, elem := range cores {
+		pCores, err := parseGroupCores(elem)
 		if err != nil {
-			// In case of an error exit the function. All metrics past this point are dependent on MSR
-			p.Log.Debugf("Error while reading msr: %v", err)
-			return
+			return nil, fmt.Errorf("failed to parse core group: %w", err)
```

Comment:
Would it provided added-value when printing which group failed?

Reply:
Yes it would! Done as part of commit [e2de6f6b](https://github.com/influxdata/telegraf/pull/14363/commits/e2de6f6b2577fe4e9ec7b420132b2750fe7e2654).